asdf/type_index.py

```diff
@@ -131,19 +131,22 @@ def _add_by_tag(self, index, name, version):
             self._type_by_name[name] = asdftype
             self._add_all_types(index, asdftype)
 
-    def _mark_used_extension(self, custom_type):
-        self._extensions_used.add(self._extension_by_cls[custom_type])
+    def _mark_used_extension(self, custom_type, serialization_context):
+        extension = self._extension_by_cls[custom_type]
+        self._extensions_used.add(extension)
+        if serialization_context is not None:
+            serialization_context.mark_extension_used(extension)
 
-    def _process_dynamic_subclass(self, custom_type):
+    def _process_dynamic_subclass(self, custom_type, serialization_context):
```

Comment:
Somewhere the concept of dynamic subclasses should be explained. In my search of the code, I didn't easily find it. Perhaps better explained in the developer documentation; more on that later

Reply:
I'm a little hazy on this myself, but I _think_ the intention of this feature was to allow serialization of subclasses that didn't exist at the time that the `TypeIndex` was constructed.  When building the type index, we inspect the `__subclasses__` attribute of each type and insert all the sub-types into a dict for fast lookup.  But that only finds subclasses that have already been imported.  Setting `handle_dynamic_subclasses = True` causes the type index to make an actual `issubclass` check against the listed types so it can catch subclasses that were imported later.  It's slower but more comprehensive.

The new API doesn't support any brand of automatically serializing sub-types.  A type must be explicitly listed in a converter in order to be serialized.  So when `TypeIndex` is dropped in 3.0 we won't have to worry about this import order issue anymore.